src/hotspot/share/nmt/regionsTree.hpp

```diff
@@ -0,0 +1,212 @@
+  SummaryDiff uncommit_region(address addr, size_t size);
+
+  using Node = VMATree::TreapNode;
+
+  class NodeHelper : public Node {
+      public:
+      inline bool is_before(VMATree::position addr) { return VMATree::PositionComparator::cmp(key(), addr) < 0; }
+      inline bool is_after(VMATree::position addr) { return VMATree::PositionComparator::cmp(key(), addr) > 0; }
+      inline bool is_region_begin() { return is_reserved_begin()|| is_committed_begin(); }
+      inline bool is_region_end(NodeHelper* node) { return is_released_begin() || (node != nullptr && node->out_flag() != out_flag()); }
+      inline bool is_end_only(NodeHelper* node) { return !is_region_begin() && is_region_end(node); }
+      inline bool is_joint(NodeHelper* node) { return is_region_begin() && is_region_end(node); }
+      inline VMATree::position position() { return key(); }
+      inline bool is_committed_begin() { return ((uint8_t)val().out.type() & (uint8_t)VMATree::StateType::Committed) >= 2; }
+      inline bool is_released_begin() { return val().out.type() == VMATree::StateType::Released; }
+      inline bool is_reserved_begin() { return ((uint8_t)val().out.type() & (uint8_t)VMATree::StateType::Reserved) == 1; }
+      inline VMATree::StateType in_state() { return val().in.type(); }
+      inline VMATree::StateType out_state() { return val().out.type(); }
+      inline size_t distance_from(NodeHelper* other) { return key() - other->key(); }
+      inline const NativeCallStack& stack(NativeCallStackStorage* ncss) { return ncss->get(val().in.stack()); }
+      inline MEMFLAGS in_flag() { return val().in.flag(); }
+      inline MEMFLAGS out_flag() { return val().out.flag(); }
+      inline void set_in_flag(MEMFLAGS flag) { val().in.set_flag(flag); }
+      inline void set_out_flag(MEMFLAGS flag) { val().out.set_flag(flag); }
+      inline void dump(outputStream* st) {
+       auto st_str = [&](int s){
+        return s == (int)VMATree::StateType::Released ? "Rl" :
+                s ==  (int)VMATree::StateType::Reserved ? "Rv" : "Cm";
+       };
+        st->print_cr("pos: " INTPTR_FORMAT " "
+                     "%s, %s <|> %s, %s",
+                     p2i((address)position()),
+                     st_str((int)val().in.type()),
+                     NMTUtil::flag_to_name(in_flag()),
+                     st_str((int)val().out.type()),
+                     NMTUtil::flag_to_name(val().out.flag())
+                     );
+      }
+    };
+
+  void dump(outputStream* st) {
+    visit_in_order([&](Node* node) {
+      NodeHelper* curr = (NodeHelper*)node;
+      curr->dump(st);
+      return true;
+    });
+  }
+
+  template<typename F> // F == bool(*)(CommittedMemoryRegion*)
+  void visit_committed_regions(const ReservedMemoryRegion* rgn, CommittedMemoryRegion* cmr, F func) {
+    size_t start = (size_t)rgn->base();
+    size_t end = (size_t)rgn->end();
+    size_t comm_size = 0;
+    size_t base = (size_t)rgn->base();
```

Comment:
Why casting these?

Reply:
Because the args to `visit_range_in_order` are of type `position` which is `size_t`. Otherwise, compiler complains about conversion from `address` to `size_t`.